### PR TITLE
Add API docs for the HasManyReference

### PR DESCRIPTION
**Pull request by bmac**

### CODE REVIEW

File: addon/-private/system/references/has-many.js [hunk 1]
```diff
@@ -26,6 +33,42 @@ HasManyReference.prototype = Object.create(Reference.prototype);
...
+
+   Example
+
+   ```javascript
+   // models/post.js
+   export default DS.Model.extend({
+     comments: DS.hasMany({ async: true })
+   });
+
+   var post = store.push({
```
Comment (pangratz) [inline addon/-private/system/references/has-many.js]: not a valid format for `store.push`

File: addon/-private/system/references/has-many.js [hunk 2]
```diff
@@ -34,10 +77,71 @@ HasManyReference.prototype.remoteType = function() {
...
```
Comment (pangratz) [inline addon/-private/system/references/has-many.js]: not a valid format for `store.push`

File: addon/-private/system/references/has-many.js [hunk 3]
```diff
@@ -46,10 +150,85 @@ HasManyReference.prototype.ids = function() {
   });
 };
 
+/**
+   The link Ember Data will use to fetch or reload this belongs-to
```
Comment (pangratz) [inline addon/-private/system/references/has-many.js]: `s/belongs-to/has-many`

File: addon/-private/system/references/has-many.js [hunk 4]
```diff
@@ -46,10 +150,85 @@ HasManyReference.prototype.ids = function() {
...
+
+   Example
+
+   ```javascript
+   // models/post.js
+   export default DS.Model.extend({
+     comments: DS.hasMany({ async: true })
+   });
+
+   var post = store.push({
```
Comment (pangratz) [inline addon/-private/system/references/has-many.js]: not a valid format for `store.push`

File: addon/-private/system/references/has-many.js [hunk 5]
```diff
@@ -34,10 +77,71 @@ HasManyReference.prototype.remoteType = function() {
...
```
Comment (pangratz) [inline addon/-private/system/references/has-many.js]: not a valid format for `store.push`

File: addon/-private/system/references/has-many.js [hunk 6]
```diff
@@ -46,10 +150,85 @@ HasManyReference.prototype.ids = function() {
...
+   @method meta
+   @return {Object} The meta information for the has-many relationship.
+*/
 HasManyReference.prototype.meta = function() {
   return this.hasManyRelationship.meta;
 };
 
+/**
+   `push` can be used to update the data in the relationship and Ember
+   Data will treat the new data as the conanical value of this
```
Comment (pangratz) [inline addon/-private/system/references/has-many.js]: typo: `s/conanical/canonical`

conanical

File: addon/-private/system/references/has-many.js [hunk 7]
```diff
@@ -46,10 +150,85 @@ HasManyReference.prototype.ids = function() {
...
+
+   Example
+
+   ```javascript
+   // models/post.js
+   export default DS.Model.extend({
+     comments: DS.hasMany({ async: true })
+   });
+
+   var post = store.push({
```
Comment (pangratz) [inline addon/-private/system/references/has-many.js]: not a valid format for `store.push`

File: addon/-private/system/references/has-many.js [hunk 8]
```diff
@@ -46,10 +150,85 @@ HasManyReference.prototype.ids = function() {
...
+         data: [{ type: 'comment', id: 1 }]
+       }
+     }
+   });
+
+   var commentsRef = post.hasMany('comments');
+
+   commentsRef.ids(); // ['1']
+
+   commentsRef.push([
```
Comment (pangratz) [inline addon/-private/system/references/has-many.js]: I am sad that this format has been added 😔, but #4398  and #4588 have been opened to address this.

Not sure if we want to advertise this in the meantime, until those PRs are in stable...

What do you think?

File: addon/-private/system/references/has-many.js [hunk 9]
```diff
@@ -121,6 +300,39 @@ HasManyReference.prototype._isLoaded = function() {
...
+
+   Example
+
+   ```javascript
+   // models/post.js
+   export default DS.Model.extend({
+     comments: DS.hasMany({ async: true })
+   });
+
+   var post = store.push({
```
Comment (pangratz) [inline addon/-private/system/references/has-many.js]: not a valid format for `store.push`

File: addon/-private/system/references/has-many.js [hunk 10]
```diff
@@ -121,6 +300,39 @@ HasManyReference.prototype._isLoaded = function() {
...
+     relationships: {
+       comments: {
+         data: [{ type: 'comment', id: 1 }]
+       }
+     }
+   });
+
+   var commentsRef = post.hasMany('comments');
+
+   commentsRef.value() === post.get('comments')
```
Comment (pangratz) [inline addon/-private/system/references/has-many.js]: This code sample is not completely accurate. We should either make the relationship sync, or resolve the `post.get("comments")` and state that the resolved value is `commentsRef.value()`.

I would prefer the promise version to indicate that references allow you synchronous access to an async relationship.

File: addon/-private/system/references/has-many.js [hunk 11]
```diff
@@ -129,6 +341,40 @@ HasManyReference.prototype.value = function() {
...
+
+   Example
+
+   ```javascript
+   // models/post.js
+   export default DS.Model.extend({
+     comments: DS.hasMany({ async: true })
+   });
+
+   var post = store.push({
```
Comment (pangratz) [inline addon/-private/system/references/has-many.js]: not a valid format for `store.push`

File: addon/-private/system/references/has-many.js [hunk 12]
```diff
@@ -137,6 +383,37 @@ HasManyReference.prototype.load = function() {
...
```
Comment (pangratz) [inline addon/-private/system/references/has-many.js]: not a valid format for `store.push`

File: addon/-private/system/references/has-many.js [hunk 13]
```diff
@@ -34,10 +77,71 @@ HasManyReference.prototype.remoteType = function() {
   return "ids";
 };
 
+/**
+   The link Ember Data will use to fetch or reload this belongs-to
```
Comment (pangratz) [inline addon/-private/system/references/has-many.js]: `s/belongs-to/has-many`

File: addon/-private/system/references/has-many.js [hunk 14]
```diff
@@ -34,10 +77,71 @@ HasManyReference.prototype.remoteType = function() {
...
+     }
+   });
+
+   var commentsRef = post.hasMany('comments');
+
+   commentsRef.link(); // '/posts/1/comments'
+   ```
+
+   @method link
+   @return {String} The link Ember Data will use to fetch or reload this belongs-to relationship.
```
Comment (pangratz) [inline addon/-private/system/references/has-many.js]: `s/belongs-to/has-many`

File: addon/-private/system/references/has-many.js (hunk 8 above)
Comment (bmac) [inline addon/-private/system/references/has-many.js]: Your call @pangratz.

File: addon/-private/system/references/has-many.js [hunk 15]
```diff
@@ -26,6 +33,44 @@ HasManyReference.prototype = Object.create(Reference.prototype);
...
 HasManyReference.prototype._super$constructor = Reference;
 
+/**
+   This returns a string that represents how the reference will be
+   looked up when it is loaded. If the relationship has a link it will
+   use the "link" otherwise it defaults to "id".
+
+   Example
+
+   ```javascript
```
Comment (locks) [inline addon/-private/system/references/has-many.js]: Can you split into

     ```app/models/post.js
     ```

and 

     ```javascript
     ```
please?

**Comment by bmac** [comment]
@locks @pangratz this pr has been rebased and updated.

File: addon/-private/system/references/has-many.js (hunk 8 above)
Comment (pangratz) [inline addon/-private/system/references/has-many.js]: I think it's ok as is for now and update the documentation once those features land.

**Comment by pangratz** [comment]
Thank much @bmac

**Pull request merged**